presto-main/src/main/java/com/facebook/presto/sql/expressions/ExpressionOptimizerManager.java

```diff
@@ -0,0 +1,152 @@
+                }
+            }
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException("Failed to load expression manager configuration", e);
+        }
+    }
+
+    private void loadExpressionOptimizerFactory(File configurationFile)
+            throws IOException
+    {
+        String name = getNameWithoutExtension(configurationFile.getName());
+        checkArgument(!isNullOrEmpty(name), "File name is empty, full path: %s", configurationFile.getAbsolutePath());
+        checkArgument(!name.equals(DEFAULT_OPTIMIZER_KEY), "Cannot name an expression optimizer instance default");
+
+        Map<String, String> properties = new HashMap<>(loadProperties(configurationFile));
+        String factoryName = properties.remove(EXPRESSION_MANAGER_FACTORY_NAME);
+        checkArgument(!isNullOrEmpty(factoryName), "%s does not contain %s", configurationFile, EXPRESSION_MANAGER_FACTORY_NAME);
+        checkArgument(expressionOptimizerFactories.containsKey(factoryName),
+                "ExpressionOptimizerFactory %s is not registered, registered factories: ", factoryName, expressionOptimizerFactories.keySet());
+        String defaultPropertyValue = properties.remove(DEFAULT_PROPERTY);
+        checkArgument(defaultPropertyValue == null || "true".equalsIgnoreCase(defaultPropertyValue) || "false".equalsIgnoreCase(defaultPropertyValue),
+                "Only true or false is allowed for %s, found: %s", DEFAULT_PROPERTY, defaultPropertyValue);
+        boolean isDefault = parseBoolean(defaultPropertyValue);
+
+        ExpressionOptimizer optimizer = expressionOptimizerFactories.get(factoryName).createOptimizer(
+                properties,
+                new ExpressionOptimizerContext(nodeManager, functionAndTypeManager, functionResolution));
+        expressionOptimizers.put(name, optimizer);
+        if (isDefault) {
+            checkArgument(defaultExpressionOptimizer.get().equals(DEFAULT_OPTIMIZER_KEY), "Default ExpressionOptimizer is already set, found: %s, attempted: %s", defaultExpressionOptimizer.get());
+            defaultExpressionOptimizer.set(name);
+        }
+    }
+
+    public void addExpressionOptimizerFactory(ExpressionOptimizerFactory expressionOptimizerFactory)
+    {
+        String name = expressionOptimizerFactory.getName();
+        checkArgument(
+                expressionOptimizerFactories.putIfAbsent(name, expressionOptimizerFactory) == null,
+                "ExpressionOptimizerFactory %s is already registered", name);
+    }
+
+    public ExpressionOptimizer getExpressionOptimizer(ConnectorSession connectorSession)
+    {
+        checkArgument(connectorSession instanceof FullConnectorSession, "connectorSession is not an instance of FullConnectorSession");
+        Session session = ((FullConnectorSession) connectorSession).getSession();
+        String expressionOptimizerInstance = getExpressionOptimizerInstance(session);
+        if (isNullOrEmpty(expressionOptimizerInstance)) {
```

Comment:
I feel like this logic is unnecessarily fragmented. Can we instead have the session property backed by a config property (like other session properties/configs) and have "default" as the default value for the config? That will also let people set desired configurations for the whole cluster.

Reply:
Good suggestion, done.